doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+`::testing::InitGoogleTest(&argc, argv)` and before `RUN_ALL_TESTS()`
+
+This are the classes involved in the testing setup:
+
+![logger testing classes](../website/images/logger_testing_classes.svg)
+
+This is the sequence diagram of the setup of the testing logger:
+
+![logger testing sequence](../website/images/logger_testing_sequence.svg)
+
+#### Environment variables
+
+The behavior of the logger can be altered via environment variables and the
+`initLogger` function. Calling this function without arguments, it will check
+whether the environment variable `IOX_LOG_LEVEL` is set or use `LogLevel::INFO`
+as log level. To have a different fallback log level, the `logLevelFromEnvOr`
+function can be used, e.g.
+
+```cpp
+iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+```
+
+If the logger shall not be altered via environment variables, `initLogger` must
+be called with the fitting log level.
+
+For the `TestingLogger` there is an additional environment variable called
+`IOX_TESTING_ALLOW_LOG`. This enables the printing of the log messages for all
+tests instead of only for failed ones.
+
+| Environment variable | Allowed values |
+|--------------|-----------|
+| IOX_LOG_LEVEL | off, fatal, error, warn, info, debug, trace |
+| IOX_TESTING_ALLOW_LOG | on, off |
+
+#### Thread local storage
+
+The `Logger` base class is using thread local storage to provide a buffer for
+each thread. In case thread local storage is not desired, the logger must be
+re-implemented via the platform abstraction.
+
+### Code example
+
+#### Using the default console logger
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+
+int main()
+{
+    iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::DEBUG));
+
+    IOX_LOG(DEBUG) << "Hello World";
+
+    return 0;
+}
+```
+
+#### Add logging support for custom types
+
+```cpp
+#include "iceoryx_hoofs/log/logging.hpp"
+#include "iceoryx_hoofs/log/logstream.hpp"
+#include <cstdint>
+
+struct MyType
+{
+    uint16_t a {0};
+    uint16_t b {0};
+};
+
+iox::log::LogStream& operator<<(iox::log::LogStream& stream, const MyType& m)
+{
+    stream << "MyType { a = " << m.a << "; b = " << m.b << " }";
+    return stream;
+}
+
+int main()
+{
+    MyType m;
+    IOX_LOG(INFO) << m;
+
+    return 0;
+}
+```
+
+#### Creating a custom logger
+
+```cpp
+#include "iceoryx_hoofs/log/logger.hpp"
+#include "iceoryx_hoofs/log/logging.hpp"
+
+class MyLogger : public iox::log::Logger
+{
+  public:
+    static void init()
+    {
+        static MyLogger myLogger;
+        iox::log::setActiveLogger(&myLogger);
+        iox::log::initLogger(iox::log::logLevelFromEnvOr(iox::log::LogLevel::INFO));
+    }
+
+  private:
+    void setupNewLogMessage(const char*, const int, const char*, iox::log::LogLevel logLevel) override
+    {
+        switch(logLevel) {
+            case iox::log::LogLevel::FATAL:
+                logString("💀: ");
```

Comment:
The example leaves out what to do with the general arguments (of type `const char*` etc.) of the call as it just ignores them. Maybe the purpose can be made more clear by using at least some of them.

Reply:
It ignores it since it is file, and function name but I should make it more clear